### ISSUE TITLE
Insufficient Funds Test Card Number Validating Incorrectly

### ISSUE THREAD

**Issue by dark-swordsman** [stripe/stripe-node#554]
Hi there. After clawing my eyes out, I finally think I understand whats going on.

I am using `react-stripe-checkout`, and I tried to trace the issue in the node_module, but I can't for the life of my figure out how it's using stripe.js to `createToken()`.

Anyways, I'm testing the error handling that we have in our API server, however, in test mode, trying the `4000 0000 0000 9995` card to throw `card_declined` for `insufficient_funds` returns the error from the stripe api server on `createToken()`. As a result, there is no token and it doesn't reach the back end.

From my understanding of the docs, this is **not** supposed to happen. All the `createToken()` function does is send the info to the stripe api and the stripe api does a validation to make sure the card exists and is valid. It doesn't actually contact the bank to verify if there are insufficient funds, as it doesn't send the amount.

We tested in production mode with a visa gift card that didn't have enough money, and it in fact **didn't** do this behavior. As expected, it made the token, the token was sent to the back end, and then when we created the payment with stripe in the back end, it threw the error as expected.

So, in conclusion, it seems to me that the test cards are **not** behaving like they are supposed to. Either that, or we really messed up in setting up our checkout process.

**Comment by ob-stripe** [comment]
Hi @dark-swordsman. This is actually expected behavior: when creating a token with [Checkout](https://stripe.com/docs/checkout), it will validate the card with the issuing bank by running a $0 or $1 authorization. (If it succeeds, the authorization is immediately released -- nothing is actually charged!)

That's why when using the `4000 0000 0000 9995` test card is immediately declined by Checkout: this test card always returns a `card_declined` error regardless of the auth's amount.

Note that if you were using [Stripe.js / Elements](https://stripe.com/docs/stripe-js) instead of Checkout to tokenize the card number, then you wouldn't run into this error. Unlike Checkout, Stripe.js only runs some basic sanity checks on the number (correct number of digits, expiry date in the future, etc.) but doesn't validate the card with the issuing bank. Rather, the card would be declined when you tried to use the token to create an actual charge.

Since this is not a problem with the stripe-node library, I'm going to close the issue. If you have follow-up questions, you should reach out to our support team at https://support.stripe.com/email or chat with our support engineers on the #stripe IRC channel on Freenode.

**Comment by francescov1** [comment]
Hi,

I am running into this same issue, but with Stripe.js / Elements. I am trying to write tests for handling errors such as insufficient_funds, but I am not able to add the card to a stripe customer. I've tried both attaching the `tok_chargeDeclinedInsufficientFunds` token and tokenizing the number `4000000000009995` and trying to attach it on a customer; both operations return the insufficient funds error **when adding the card to the customer**, not when making a payment.

**Comment by remi-stripe** [comment]
To attach a card that will fail when you charge (but not when you attach it) you need to use `tok_chargeCustomerFail` instead

**Comment by francescov1** [comment]
This only throws the generic error though, is there no way to test specific errors such as `insufficient_funds`? This would be really useful for us.

Additionally, I haven't found clear documentation that defines which errors can be thrown when (when adding a card vs. when making a payment). It was mentioned above that with Stripe.js / Elements there's no validation with bank happening, but with `insufficient_funds` being thrown when adding a card in test mode, this is quite unclear.

**Comment by remi-stripe** [comment]
To throw that specific error you would use `tok_chargeDeclinedInsufficientFunds` when charging the card (and not when attaching it).

I'd recommend talking to our support team directly if you have more questions about your integration: https://support.stripe.com/contact
